### ISSUE TITLE
Rule proposal: Enforce presence of optional parameters in override method signatures

### ISSUE THREAD

**Issue by MGibson1** [typescript-eslint/typescript-eslint#8049]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Enforce presence of optional parameters in override method signatures.

### Fail Cases

```TypeScript
class Foo {
  foo(s?: string) {
    console.log(`foo: ${s}`);
  }
}

class Bar extends Foo {
  override foo() {
    console.log("bar");
  }
}
```

### Pass Cases

```TypeScript
class Foo {
  foo(s?: string) {
    console.log(`foo: ${s}`);
  }
}

class Bar extends Foo {
  override foo(s?: string) {
    console.log("bar");
  }
}
```

### Additional Info

Typescript requires that the method signatures are compatible, but this does not require explicit handling of all parameters, making it somewhat easy to drop optional parameters in overrides. It should suffice to count the number of parameters in the parent implementation and require the same count in the override. Typescript will enforce the rest.

**Comment by Josh-Cena** [comment]
What bugs does this catch? Usually when you turn on `no-unused-vars` you will be prompted to remove unused parameters anyway.

**Comment by MGibson1** [comment]
I hadn't though about collisions with unused vars. You're right.

[This is the bug](https://github.com/bitwarden/clients/pull/7176) that prompted the proposal. I must admit that I think that the optional parameter is a bit of a smell, but visibility into oddities in the override would have helped in this case.

**Comment by bradzacher** [comment]
Overrides check assignability of the child implementation to the parent type.
Any function with fewer parameters (eg `() => void`) is assignable to a function with more parameters (eg `(a?: string) => void` AND `(a: string | undefined) => void`) 

So from a type-system perspective it makes complete sense that your implementation type-checked fine.

From an implementation perspective - it also makes sense - there are many valid cases where you specifically want to ignore the extra parameters in a child class.
I don't think that you can specifically lint for this case because of this - there are many cases where the rule will false positive and your only recourse would be to either (a) add a disable comment to ignore the rule or (b) add the missing parameters.
However for (b) this would (as josh mentioned) trigger `no-unused-vars` (which is turned on in pretty well every codebase!) so you'd either need disable comments for those or you'd need to name them in specific ways to work around the rule (if your config allows this).

I definitely understand the sentiment here and why you might want such a rule in certain cases - however I don't think such a rule is going to be broadly applicable given the above. We'd certainly never be able to include it in our recommended or strict configs for this reason!

For this reason I am going to reject this proposal from our project.
We encourage you to explore building on top of our tooling though! Please feel free to use our tooling and either build an internal lint rule or publish a 3rd party plugin.
